saltstack/base/salt/userdata-secrets/bin/retrieve-userdata-secrets.sh

```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+set -e +x
+
+# This function implements basically the same exponential backoff strategy as the aws cli in standard retry mode
+retry_until_non_placeholder_value() {
+  local BASE_FACTOR=2
+  local MAX_BACKOFF=20
+  local MAX_ATTEMPT=15
+
+  local total_wait_time=0
+  local timeout=1
+  local attempt=1
+
+  SECRET_STRING="$("$@")"
+  while [[ "$SECRET_STRING" == "PLACEHOLDER" ]]; do
+    if (( attempt > MAX_ATTEMPT )); then
+      echo "Maximum number of retry attempts reached while waiting for userdata secret to update! Total wait time was $total_wait_time seconds. Exiting..."
+      exit 1
+    fi
+
+    echo "Userdata secret still has PLACEHOLDER value. Waiting $timeout seconds until the next attempt to retrieve the secret."
+    sleep $timeout
+
+    total_wait_time=$(( total_wait_time + timeout ))
+    attempt=$(( attempt + 1))
+    timeout=$(( timeout * BASE_FACTOR > MAX_BACKOFF ? MAX_BACKOFF : timeout * BASE_FACTOR ))
+
+    SECRET_STRING="$("$@")"
+  done
+  echo "Userdata secret's value was retrieved after waiting $total_wait_time seconds."
+}
+
+# Intermittent issues like throttling of AWS APIs are handled by the built in retry mechanism of the aws cli.
+# Waiting for CB to update the secret's value to the actual secrets is handled by the function above.
+main() {
+  if [[ "$CLOUD_PLATFORM" == "AWS" ]]; then
+    # See: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-retries.html
+    # aws cli with standard retry mode uses an exponential backoff with a base factor of 2 and a maximum backoff time of 20 seconds
+    aws_retry_mode=standard
+    # AWS_MAX_ATTEMPTS=15 would mean a maximum wait time of approximately 4 minutes per aws cli command
+    aws_max_attempts=15
+
+    echo "Retrieving userdata secrets..."
+    AWS_RETRY_MODE="$aws_retry_mode" \
+    AWS_MAX_ATTEMPTS="$aws_max_attempts" \
+    retry_until_non_placeholder_value \
+    aws secretsmanager get-secret-value \
```

Comment:
While `retry_until_non_placeholder_value()` inherits the changes of `AWS_RETRY_MODE` and `AWS_MAX_ATTEMPTS`, they are not exposed to the child process started by `SECRET_STRING="$("$@")"`, unless these environment variables are marked for `export`.

I suggest either

- Turning these variable assignments into explicit `export` commands executed on their own (i.e. dropping the backslash at the end of lines 45 and 46); or
- Reording the lines, making `retry_until_non_placeholder_value` the first one so that the variable assignments become part of its arguments.